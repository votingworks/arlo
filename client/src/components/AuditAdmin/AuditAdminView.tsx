import React from 'react'
import { useParams, Redirect } from 'react-router-dom'
import { Spinner, H3, Intent } from '@blueprintjs/core'
import { useQueryClient } from 'react-query'
import {
  useRounds,
  isDrawSampleComplete,
  drawSampleError,
  isDrawingSample,
  useStartNextRound,
  useUndoRoundStart,
  ISampleSizes,
  roundsQueryKey,
} from './useRoundsAuditAdmin'
import { useJurisdictions, jurisdictionsQueryKey } from '../useJurisdictions'
import { useContests } from '../useContests'
import { useAuditSettings } from '../useAuditSettings'
import { Wrapper, Inner } from '../Atoms/Wrapper'
import { AuditAdminStatusBox } from '../Atoms/StatusBox'
import { RefreshTag } from '../Atoms/RefreshTag'
import Setup from './Setup/Setup'
import Progress from './Progress/Progress'

interface IParams {
  electionId: string
  view: 'setup' | 'progress' | ''
}

const AuditAdminView: React.FC = () => {
  const { electionId, view } = useParams<IParams>()

  const roundsQuery = useRounds(electionId, {
    refetchInterval: rounds =>
      rounds && isDrawingSample(rounds) ? 1000 : false,
  })
  const startNextRoundMutation = useStartNextRound(electionId)
  const undoRoundStartMutation = useUndoRoundStart(electionId)

  const jurisdictionsQuery = useJurisdictions(electionId)
  const contestsQuery = useContests(electionId)
  const auditSettingsQuery = useAuditSettings(electionId)
  const queryClient = useQueryClient()

  if (
    !jurisdictionsQuery.isSuccess ||
    !contestsQuery.isSuccess ||
    !roundsQuery.isSuccess ||
    !auditSettingsQuery.isSuccess
  )
    return null // Still loading

  const jurisdictions = jurisdictionsQuery.data
  const contests = contestsQuery.data
  const rounds = roundsQuery.data
  const auditSettings = auditSettingsQuery.data

  if (rounds.length > 0 && !isDrawSampleComplete(rounds)) {
    return (
      <Wrapper>
        <Inner>
          <div
            style={{
              display: 'flex',
              flexDirection: 'column',
              alignItems: 'center',
              width: '100%',
              marginTop: '100px',
            }}
          >
            <div style={{ marginBottom: '20px' }}>
              <Spinner size={Spinner.SIZE_LARGE} intent={Intent.PRIMARY} />
            </div>
            <H3>Drawing a random sample of ballots...</H3>
            <p>For large elections, this can take a couple of minutes.</p>
          </div>
        </Inner>
      </Wrapper>
    )
  }

  const startNextRound = async (sampleSizes: ISampleSizes) => {
    const nextRoundNum =
      rounds.length === 0 ? 1 : rounds[rounds.length - 1].roundNum + 1
    await startNextRoundMutation.mutateAsync({
      sampleSizes,
      roundNum: nextRoundNum,
    })
    return true
  }

  const undoRoundStart = async () => {
    const currentRoundId = rounds[rounds.length - 1].id
    await undoRoundStartMutation.mutateAsync(currentRoundId)
    return true
  }

  switch (view) {
    case 'setup':
      return (
        <Wrapper>
          <AuditAdminStatusBox
            rounds={rounds}
            startNextRound={startNextRound}
            undoRoundStart={undoRoundStart}
            jurisdictions={jurisdictions}
            contests={contests}
            auditSettings={auditSettings}
          />
          <Setup
            electionId={electionId}
            auditSettings={auditSettings}
            startNextRound={startNextRound}
            isAuditStarted={rounds.length > 0}
          />
        </Wrapper>
      )
    case 'progress':
      return (
        <Wrapper>
          <AuditAdminStatusBox
            rounds={rounds}
            startNextRound={startNextRound}
            undoRoundStart={undoRoundStart}
            jurisdictions={jurisdictions}
            contests={contests}
            auditSettings={auditSettings}
          >
            <RefreshTag
              refresh={() => {
                queryClient.invalidateQueries(roundsQueryKey(electionId))
                queryClient.invalidateQueries(jurisdictionsQueryKey(electionId))
              }}
            />
          </AuditAdminStatusBox>
          {!drawSampleError(rounds) && (
            <Inner>
              <Progress
                jurisdictions={jurisdictions}
                auditSettings={auditSettings}
                round={rounds.length > 0 ? rounds[rounds.length - 1] : null}
              />
            </Inner>
          )}
        </Wrapper>
      )
    default:
      return (
        <Redirect
          to={
            rounds.length > 0
              ? `/election/${electionId}/progress`
              : `/election/${electionId}/setup`
          }
        />
      )
  }
}

export default AuditAdminView
