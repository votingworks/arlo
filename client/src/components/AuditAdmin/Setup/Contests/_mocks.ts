export const contestsInputMocks = {
  inputs: [
    { key: 'Contest Name', value: 'Contest Name' },
    { key: 'Name of Candidate/Choice 1', value: 'Choice One' },
    { key: 'Name of Candidate/Choice 2', value: 'Choice Two' },
    { key: 'Votes for Candidate/Choice 1', value: '10' },
    { key: 'Votes for Candidate/Choice 2', value: '20' },
    { key: 'Total Ballot Cards Cast for Contest', value: '30' },
  ],
  errorInputs: [
    { key: 'Contest Name', value: '', error: 'Required' },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '',
      error:
        'Must be greater than or equal to the sum of votes for each candidate/choice',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Total Ballot Cards Cast for Contest',
      value: '0.5',
      error: 'Must be an integer',
    },
    { key: 'Name of Candidate/Choice 1', value: '', error: 'Required' },
    { key: 'Name of Candidate/Choice 2', value: '', error: 'Required' },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 1',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '',
      error: 'Required',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'Votes for Candidate/Choice 2',
      value: '0.5',
      error: 'Must be an integer',
    },
  ],
}

export default contestsInputMocks
