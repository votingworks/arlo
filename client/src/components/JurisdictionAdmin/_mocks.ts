import { IBatches } from './useBatchResults'
import { IRound } from '../AuditAdmin/useRoundsAuditAdmin'
import { FileProcessingStatus } from '../useCSV'
import {
  IFullHandTallyBatchResults,
  IFullHandTallyBatchResult,
} from './useFullHandTallyResults'
import { mocksOfType } from '../testUtilities'

export interface INullResultValues {
  [contestId: string]: {
    [choiceId: string]: null | string | number
  }
}

export const roundMocks = mocksOfType<IRound>()({
  incomplete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: null,
    isAuditComplete: false,
    needsFullHandTally: false,
    isFullHandTally: false,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
  complete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: '2020-09-14T17:35:19.482Z',
    needsFullHandTally: false,
    isAuditComplete: true,
    isFullHandTally: false,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
  fullHandTallyIncomplete: {
    id: 'round-1',
    roundNum: 1,
    startedAt: '2020-09-14T17:35:19.482Z',
    endedAt: null,
    isAuditComplete: false,
    needsFullHandTally: true,
    isFullHandTally: true,
    drawSampleTask: {
      status: FileProcessingStatus.PROCESSED,
      startedAt: '2020-09-14T17:35:19.482Z',
      completedAt: '2020-09-14T17:36:19.482Z',
      error: null,
    },
  },
})

export const resultsMocks = mocksOfType<INullResultValues>()({
  emptyInitial: {
    'contest-id-1': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'contest-id-2': {
      'choice-id-3': null,
      'choice-id-4': null,
    },
  },
  complete: {
    'contest-id-1': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'contest-id-2': {
      'choice-id-3': 1,
      'choice-id-4': 2,
    },
  },
})

export const batchResultsMocks = mocksOfType<INullResultValues>()({
  empty: {
    'batch-1': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'batch-2': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
    'batch-3': {
      'choice-id-1': null,
      'choice-id-2': null,
    },
  },
  complete: {
    'batch-1': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'batch-2': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
    'batch-3': {
      'choice-id-1': 1,
      'choice-id-2': 2,
    },
  },
})

export const batchesMocks = mocksOfType<IBatches>()({
  emptyInitial: {
    batches: [
      {
        id: 'batch-1',
        name: 'Batch One',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
      {
        id: 'batch-2',
        name: 'Batch Two',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
      {
        id: 'batch-3',
        name: 'Batch Three',
        numBallots: 100,
        auditBoard: null,
        resultTallySheets: [],
      },
    ],
    resultsFinalizedAt: null,
  },
  complete: {
    batches: [
      {
        id: 'batch-1',
        name: 'Batch One',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 1,
              'choice-id-2': 2,
            },
          },
        ],
      },
      {
        id: 'batch-2',
        name: 'Batch Two',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 0,
              'choice-id-2': 10,
            },
          },
        ],
      },
      {
        id: 'batch-3',
        name: 'Batch Three',
        numBallots: 100,
        auditBoard: {
          id: 'ab-1',
          name: 'Audit Board One',
        },
        resultTallySheets: [
          {
            name: 'Tally Sheet #1',
            results: {
              'choice-id-1': 2000,
              'choice-id-2': 20,
            },
          },
        ],
      },
    ],
    resultsFinalizedAt: '2020-09-14T17:35:19.482Z',
  },
})

export const fullHandTallyBatchResultMock = mocksOfType<
  IFullHandTallyBatchResults
>()({
  empty: {
    finalizedAt: '',
    results: [],
  },
  complete: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch1',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
  completeWithMultipleBatch: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch1',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 5,
          'choice-id-2': 15,
        },
      },
      {
        batchName: 'Batch2',
        batchType: 'Provisional',
        choiceResults: {
          'choice-id-1': 5,
          'choice-id-2': 5,
        },
      },
    ],
  },
  updated: {
    finalizedAt: '',
    results: [
      {
        batchName: 'Batch12',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
  finalized: {
    finalizedAt: '2021-04-13T15:01:00.383031+00:00',
    results: [
      {
        batchName: 'Batch12',
        batchType: 'Other',
        choiceResults: {
          'choice-id-1': 10,
          'choice-id-2': 20,
        },
      },
    ],
  },
})

export const fullHandTallyBatchResultsMock = mocksOfType<
  IFullHandTallyBatchResult
>()({
  empty: {
    batchName: '',
    batchType: '',
    choiceResults: {},
  },
  complete: {
    batchName: 'Batch1',
    batchType: 'Other',
    choiceResults: {
      'choice-id-1': 10,
      'choice-id-2': 20,
    },
  },
  updated: {
    batchName: 'Batch12',
    batchType: 'Other',
    choiceResults: {
      'choice-id-1': 10,
      'choice-id-2': 20,
    },
  },
})
