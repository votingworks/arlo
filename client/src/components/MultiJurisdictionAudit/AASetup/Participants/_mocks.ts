import { readFileSync } from 'fs'
import { join } from 'path'

const jurisdictionFile = new File(
  [
    readFileSync(
      join(
        __dirname,
        '../../../../../public/sample_jurisdiction_filesheet.csv'
      ),
      'utf8'
    ),
  ],
  'jurisdictions.csv',
  { type: 'text/csv' }
)

export default jurisdictionFile
