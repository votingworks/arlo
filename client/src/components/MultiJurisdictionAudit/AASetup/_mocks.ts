import { ElementType } from '../../../types'
import { setupStages, stageTitles } from './index'
import { ISidebarMenuItem } from '../../Atoms/Sidebar'

const relativeStages = (
  stage: ElementType<typeof setupStages>,
  state: ISidebarMenuItem['state'] = 'live'
): {
  prevStage: ISidebarMenuItem
  nextStage: ISidebarMenuItem
  menuItems: ISidebarMenuItem[]
} => {
  const prevId = setupStages[setupStages.indexOf(stage) - 1]
  const prevStage: ISidebarMenuItem = {
    id: prevId,
    title: stageTitles[prevId],
    active: false,
    activate: jest.fn(),
    state,
  }
  const nextId = setupStages[setupStages.indexOf(stage) + 1]
  const nextStage: ISidebarMenuItem = {
    id: nextId,
    title: stageTitles[nextId],
    active: false,
    activate: jest.fn(),
    state,
  }
  const menuItems = setupStages.map((s: ElementType<typeof setupStages>) => ({
    id: s,
    title: stageTitles[s],
    active: s === stage,
    activate: jest.fn(),
    state,
  }))
  return { prevStage, nextStage, menuItems }
}

export default relativeStages
