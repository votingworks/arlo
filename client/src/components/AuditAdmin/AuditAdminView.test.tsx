import React from 'react'
import { waitFor, fireEvent, screen, within } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Route, RouteProps } from 'react-router-dom'
import { QueryClientProvider } from 'react-query'
import AuditAdminView from './AuditAdminView'
import {
  jurisdictionFileMocks,
  standardizedContestsFileMocks,
  auditSettings,
  roundMocks,
  manifestFile,
  manifestMocks,
  jurisdictionMocks,
} from './useSetupMenuItems/_mocks'
import { withMockFetch, renderWithRouter } from '../testUtilities'
import AuthDataProvider, { useAuthDataContext } from '../UserContext'
import getJurisdictionFileStatus from './useSetupMenuItems/getJurisdictionFileStatus'
import getRoundStatus from './useSetupMenuItems/getRoundStatus'
import { aaApiCalls, jaApiCalls } from '../_mocks'
import {
  jurisdictionFile,
  jurisdictionErrorFile,
  standardizedContestsFile,
} from './Setup/Participants/_mocks'
import { queryClient } from '../../App'

const getJurisdictionFileStatusMock = getJurisdictionFileStatus as jest.Mock
const getRoundStatusMock = getRoundStatus as jest.Mock

jest.mock('./useSetupMenuItems/getJurisdictionFileStatus')
jest.mock('./useSetupMenuItems/getRoundStatus')
getJurisdictionFileStatusMock.mockReturnValue('PROCESSED')
getRoundStatusMock.mockReturnValue(false)

jest.mock('axios')

// AuditAdminView will only be rendered once the user is logged in, so
// we simulate that.
const AuditAdminViewWithAuth = (props: RouteProps) => {
  const auth = useAuthDataContext()
  return auth ? <AuditAdminView {...props} /> : null
}

const render = (view = 'setup') =>
  renderWithRouter(
    <QueryClientProvider client={queryClient}>
      <AuthDataProvider>
        <Route
          path="/election/:electionId/:view?"
          render={routeProps => <AuditAdminViewWithAuth {...routeProps} />}
        />
      </AuthDataProvider>
    </QueryClientProvider>,
    { route: `/election/1/${view}` }
  )

describe('AA setup flow', () => {
  const loadEach = [
    aaApiCalls.getRounds([]),
    aaApiCalls.getJurisdictions,
    aaApiCalls.getContests,
    aaApiCalls.getSettings(auditSettings.all),
  ]

  it('sidebar changes stages', async () => {
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadEach,
      ...loadEach,
      aaApiCalls.getSettings(auditSettings.all),
      aaApiCalls.getJurisdictionFile,
      ...loadEach,
      aaApiCalls.getSettings(auditSettings.all),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { queryAllByText, getByText } = render()

      await waitFor(() => {
        expect(queryAllByText('Participants')).toHaveLength(2)
      })

      fireEvent.click(getByText('Audit Settings') as Element, { bubbles: true })

      await waitFor(() => {
        expect(queryAllByText('Audit Settings')).toHaveLength(2)
      })
    })
  })

  it('renders sidebar when authenticated on /setup', async () => {
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadEach,
      ...loadEach,
      aaApiCalls.getSettings(auditSettings.all),
      aaApiCalls.getJurisdictionFile,
    ]
    await withMockFetch(expectedCalls, async () => {
      const { container, queryAllByText } = render()

      await waitFor(() => {
        expect(queryAllByText('Participants')).toHaveLength(2)
      })
      expect(container).toMatchSnapshot()
    })
  })

  it('get empty jurisdiction file intiially', async () => {
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadEach,
      ...loadEach,
      aaApiCalls.getSettings(auditSettings.blank),
      aaApiCalls.getJurisdictionFileWithResponse(jurisdictionFileMocks.empty),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { queryAllByText } = render()

      await waitFor(() => {
        expect(queryAllByText('Participants')).toHaveLength(2)
        screen.getByLabelText('Select a file...')
      })
    })
  })

  it('get jurisdisction file get if exists', async () => {
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadEach,
      ...loadEach,
      aaApiCalls.getSettings(auditSettings.all),
      aaApiCalls.getJurisdictionFile,
    ]
    await withMockFetch(expectedCalls, async () => {
      const { queryByText, queryAllByText } = render()

      await waitFor(() => {
        expect(queryAllByText('Participants')).toHaveLength(2)
        expect(queryByText(/Current file/))
      })
    })
  })

  it('jurisdisction file upload success', async () => {
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadEach,
      ...loadEach,
      aaApiCalls.getSettings(auditSettings.blank),
      aaApiCalls.getJurisdictionFileWithResponse(jurisdictionFileMocks.empty),
      aaApiCalls.putJurisdictionFile,
      aaApiCalls.getJurisdictionFileWithResponse(
        jurisdictionFileMocks.processed
      ),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { queryAllByText } = render()

      await waitFor(() => {
        expect(queryAllByText('Participants')).toHaveLength(2)
      })
      const jurisdisctionInput = screen.getByLabelText('Select a file...')
      const jurisdictionButton = screen.getByRole('button', {
        name: 'Upload File',
      })
      userEvent.click(jurisdictionButton)
      await screen.findByText('You must upload a file')

      userEvent.upload(jurisdisctionInput, jurisdictionFile)
      userEvent.click(jurisdictionButton)
      await screen.findByText(/Uploaded/)
    })
  })

  it('jurisdisction file upload with error', async () => {
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadEach,
      ...loadEach,
      aaApiCalls.getSettings(auditSettings.blank),
      aaApiCalls.getJurisdictionFileWithResponse(jurisdictionFileMocks.empty),
      aaApiCalls.putJurisdictionErrorFile,
      aaApiCalls.getJurisdictionFileWithResponse(jurisdictionFileMocks.errored),
    ]
    await withMockFetch(expectedCalls, async () => {
      const { queryAllByText } = render()

      await waitFor(() => {
        expect(queryAllByText('Participants')).toHaveLength(2)
      })
      const jurisdisctionInput = screen.getByLabelText('Select a file...')
      const jurisdictionButton = screen.getByRole('button', {
        name: 'Upload File',
      })
      userEvent.click(jurisdictionButton)
      await screen.findByText('You must upload a file')

      userEvent.upload(jurisdisctionInput, jurisdictionErrorFile)
      userEvent.click(jurisdictionButton)
      await screen.findByText('Invalid CSV')
    })
  })

  it('standardized contests file upload success', async () => {
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadEach,
      ...loadEach,
      aaApiCalls.getSettings(auditSettings.blankBallotComparison),
      aaApiCalls.getJurisdictionFile,
      aaApiCalls.getStandardizedContestsFileWithResponse(
        standardizedContestsFileMocks.empty
      ),
      aaApiCalls.putStandardizedContestsFile,
      aaApiCalls.getStandardizedContestsFileWithResponse(
        standardizedContestsFileMocks.processed
      ),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()

      // check file upload of jurisdiction
      await screen.findByText(/Uploaded/)
      const standardizedContestInput = screen.getByLabelText('Select a file...')
      const standardizedContestButton = screen.getByRole('button', {
        name: 'Upload File',
      })
      userEvent.click(standardizedContestButton)
      await screen.findByText('You must upload a file')

      userEvent.upload(standardizedContestInput, standardizedContestsFile)
      userEvent.click(standardizedContestButton)
      await screen.findByText(/Uploaded/)
    })
  })

  it('redirects to /progress after audit is launched', async () => {
    const loadAfterLaunch = [
      aaApiCalls.getRounds(roundMocks.singleIncomplete),
      aaApiCalls.getJurisdictions,
      aaApiCalls.getContests,
      aaApiCalls.getSettings(auditSettings.all),
    ]
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadAfterLaunch,
      ...loadAfterLaunch,
      ...loadAfterLaunch,
      aaApiCalls.getMapData,
    ]
    await withMockFetch(expectedCalls, async () => {
      const { history } = render('')
      await waitFor(() => {
        expect(history.location.pathname).toEqual('/election/1/progress')
      })
    })
  })

  it('shows an error and undo button if drawing the sample fails', async () => {
    const loadAfterLaunch = [
      aaApiCalls.getRounds(roundMocks.drawSampleErrored),
      aaApiCalls.getJurisdictions,
      aaApiCalls.getContests,
      aaApiCalls.getSettings(auditSettings.all),
    ]
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadAfterLaunch,
      ...loadAfterLaunch,
      aaApiCalls.getSettings(auditSettings.all),
      aaApiCalls.getJurisdictionFile,
      {
        url: '/api/election/1/round/round-1',
        options: { method: 'DELETE' },
        response: { status: 'ok' },
      },
      aaApiCalls.getRounds(roundMocks.empty),
    ]
    await withMockFetch(expectedCalls, async () => {
      render()
      await screen.findByRole('heading', {
        name: 'Arlo could not draw the sample',
      })
      screen.getByText(
        'Please contact our support team for help resolving this issue.'
      )
      screen.getByText('Error: something went wrong')

      userEvent.click(screen.getByRole('button', { name: 'Undo Audit Launch' }))
      await screen.findByText('The audit has not started.')
    })
  })

  it('reloads jurisdiction progress after file upload', async () => {
    const expectedCalls = [
      aaApiCalls.getUser,
      ...loadEach,
      ...loadEach,
      aaApiCalls.getMapData,
      jaApiCalls.getBallotManifestFile(manifestMocks.empty),
      jaApiCalls.putManifest,
      jaApiCalls.getBallotManifestFile(manifestMocks.processed),
      {
        ...aaApiCalls.getJurisdictions,
        response: { jurisdictions: jurisdictionMocks.allManifests },
      },
    ]
    await withMockFetch(expectedCalls, async () => {
      const { container } = render('progress')

      await waitFor(() => {
        expect(container.querySelectorAll('.bp3-spinner').length).toBe(0)
      })

      screen.getByText('Audit Progress')
      let rows = screen.getAllByRole('row')
      let row1 = within(rows[1]).getAllByRole('cell')
      expect(row1[0]).toHaveTextContent('Jurisdiction One')
      within(row1[1]).getByText('No manifest uploaded')

      // Click on a jurisdiction name to open the detail modal
      userEvent.click(screen.getByRole('button', { name: 'Jurisdiction One' }))
      const modal = screen
        .getByRole('heading', { name: 'Jurisdiction One' })
        .closest('div.bp3-dialog')! as HTMLElement
      await within(modal).findByText('Ballot Manifest')

      // Upload a manifest
      userEvent.upload(
        within(modal).getByLabelText('Select a file...'),
        manifestFile
      )
      await within(modal).findByText('manifest.csv')
      userEvent.click(within(modal).getByRole('button', { name: /Upload/ }))
      await within(modal).findByText('Uploaded')

      // Close the detail modal
      userEvent.click(screen.getByRole('button', { name: 'Close' }))

      // Jurisdiction table should be updated
      rows = screen.getAllByRole('row')
      row1 = within(rows[1]).getAllByRole('cell')
      within(row1[1]).getByText('Manifest uploaded')
    })
  })
})
