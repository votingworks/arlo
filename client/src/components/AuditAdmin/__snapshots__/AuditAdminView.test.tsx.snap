// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AA setup flow renders sidebar when authenticated on /setup 1`] = `
<div>
  <div
    class="sc-kEYyzF gOPOqs"
  >
    <div
      class="bp3-callout sc-hSdWYo gMDgEE"
    >
      <div
        class="sc-kkGfuU gzyJPg"
      >
        <div
          class="text"
        >
          <h3
            class="bp3-heading sc-iAyFgw fUjqOI"
          >
            Test Audit
          </h3>
          <h4
            class="bp3-heading"
          >
            The audit has not started.
          </h4>
          <p>
            Audit setup is complete.
          </p>
          <p>
            0 of 2 jurisdictions have completed file uploads.
          </p>
          <span
            class="bp3-tag sc-jWBwVP fvZRn"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              Will refresh in 5 minutes
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="sc-kkGfuU gzyJPg"
    >
      <div
        class="sc-cvbbAY vOHMb"
      >
        <h2
          class="bp3-heading sc-ifAKCX cmZyyF"
        >
          Audit Setup
        </h2>
        <ul
          class="bp3-menu"
        >
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-active bp3-intent-primary bp3-popover-dismiss"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                Participants
              </div>
            </a>
          </li>
          <li
            class="bp3-menu-divider"
          />
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-disabled"
              tabindex="-1"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                Target Contests
              </div>
              <span
                class="bp3-menu-item-label"
              >
                <div
                  class="bp3-spinner"
                >
                  <div
                    class="bp3-spinner-animation"
                  >
                    <svg
                      height="20"
                      stroke-width="16.00"
                      viewBox="-3.00 -3.00 106.00 106.00"
                      width="20"
                    >
                      <path
                        class="bp3-spinner-track"
                        d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                      />
                      <path
                        class="bp3-spinner-head"
                        d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                        pathLength="280"
                        stroke-dasharray="280 280"
                        stroke-dashoffset="210"
                      />
                    </svg>
                  </div>
                </div>
              </span>
            </a>
          </li>
          <li
            class="bp3-menu-divider"
          />
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-disabled"
              tabindex="-1"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                Opportunistic Contests
              </div>
              <span
                class="bp3-menu-item-label"
              >
                <div
                  class="bp3-spinner"
                >
                  <div
                    class="bp3-spinner-animation"
                  >
                    <svg
                      height="20"
                      stroke-width="16.00"
                      viewBox="-3.00 -3.00 106.00 106.00"
                      width="20"
                    >
                      <path
                        class="bp3-spinner-track"
                        d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                      />
                      <path
                        class="bp3-spinner-head"
                        d="M 50,50 m 0,-45 a 45,45 0 1 1 0,90 a 45,45 0 1 1 0,-90"
                        pathLength="280"
                        stroke-dasharray="280 280"
                        stroke-dashoffset="210"
                      />
                    </svg>
                  </div>
                </div>
              </span>
            </a>
          </li>
          <li
            class="bp3-menu-divider"
          />
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                Audit Settings
              </div>
            </a>
          </li>
          <li
            class="bp3-menu-divider"
          />
          <li
            class=""
          >
            <a
              class="bp3-menu-item bp3-popover-dismiss"
            >
              <div
                class="bp3-text-overflow-ellipsis bp3-fill"
              >
                Review & Launch
              </div>
            </a>
          </li>
        </ul>
      </div>
      <div
        class="sc-EHOje bjGJmd"
      >
        <h2
          class="bp3-heading sc-ifAKCX cmZyyF"
        >
          Participants
        </h2>
        <form>
          <div
            class="sc-EHOje bjGJmd"
          >
            <div>
              <div
                class="sc-gzVnrw kvvweo"
              >
                Click "Browse" to choose the appropriate file from your computer. This file should be a comma-separated list of all the jurisdictions participating in the audit, plus email addresses for audit administrators in each participating jurisdiction.
                <br />
                <br />
                <a
                  href="/sample_jurisdiction_filesheet.csv"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  (Click here to view a sample file in the correct format.)
                </a>
              </div>
            </div>
            <div>
              <p>
                <strong>
                  Current file:
                </strong>
                 
                file name
              </p>
              <p
                class="sc-dnqmqq jvyRYy"
              >
                Uploaded at
                 
                12/4/2020, 2:32:15 AM
                .
              </p>
              <div>
                <button
                  class="bp3-button sc-bxivhb WkpPV"
                  type="button"
                >
                  <span
                    class="bp3-button-text"
                  >
                    Replace File
                  </span>
                </button>
              </div>
            </div>
          </div>
        </form>
        <div
          class="sc-bdVaJa kcYonD"
        >
          <button
            class="bp3-button bp3-intent-primary sc-bxivhb WkpPV"
            type="submit"
          >
            <span
              class="bp3-button-text"
            >
              Next
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
